hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/AbstractMapSetMessageTask.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.protocol.task.map;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.instance.Node;
+import com.hazelcast.map.impl.MapContainer;
+import com.hazelcast.map.impl.MapService;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.security.permission.ActionConstants;
+import com.hazelcast.security.permission.MapPermission;
+
+import java.security.Permission;
+
+public abstract class AbstractMapSetMessageTask<P> extends AbstractMapPartitionMessageTask<P> {
```

Comment:
can't we have same functionality by introducing a flag? when operation is `put` set it to true, otherwise false. By using this flag, we can get rid of this class and can use `AbstractMapPutMessageTask` instead?

Reply:
In this case we bring details of sublclasses to the AbstractMapPutMessageTask's level that there are two groups map put tasks, one with flag=true, other one with flag=false. And, we will have to force all subclasses to choose the correct flag. Also, potentially, in the future, we might want to separate stats calculation for "replace" operation too (now it is counted as "put"), in that case a boolean flag won't be enough.
So, summarizing all above, I think the logic separation by one more abstracted class is preferable here. What do you think?

Reply:
ok it makes sense.